scheduling/internal/decisions/explanation/explainer.go

```diff
@@ -6,72 +6,879 @@ package explanation
+	lastDecision := &v1alpha1.Decision{}
+	if err := e.Get(ctx, client.ObjectKey{
+		Namespace: lastDecisionRef.Namespace,
+		Name:      lastDecisionRef.Name,
+	}, lastDecision); err != nil {
+		logger := log.FromContext(ctx)
+		if apierrors.IsNotFound(err) {
+			logger.Info("History decision not found, skipping history comparison",
+				"decision", lastDecisionRef.Name,
+				"namespace", lastDecisionRef.Namespace,
+				"uid", lastDecisionRef.UID)
+			return "", nil // Skip history comparison instead of failing
+		}
+		// For other errors, still fail
+		return "", err
+	}
+
+	lastTarget := "(n/a)"
+	if lastDecision.Status.Result != nil && lastDecision.Status.Result.TargetHost != nil {
+		lastTarget = *lastDecision.Status.Result.TargetHost
+	}
+
+	newTarget := "(n/a)"
+	if decision.Status.Result != nil && decision.Status.Result.TargetHost != nil {
+		newTarget = *decision.Status.Result.TargetHost
+	}
+
+	return fmt.Sprintf("Previous target host was '%s', now it's '%s'.", lastTarget, newTarget), nil
+}
+
+// getResourceType returns a human-readable resource type.
+func (e *Explainer) getResourceType(decisionType v1alpha1.DecisionType) string {
+	switch decisionType {
 	case v1alpha1.DecisionTypeNovaServer:
-		resourceType = "nova server"
+		return "nova server"
 	case v1alpha1.DecisionTypeManilaShare:
-		resourceType = "manila share"
+		return "manila share"
 	case v1alpha1.DecisionTypeCinderVolume:
-		resourceType = "cinder volume"
+		return "cinder volume"
 	case v1alpha1.DecisionTypeIroncoreMachine:
-		resourceType = "ironcore machine"
+		return "ironcore machine"
 	default:
-		resourceType = "resource"
+		return "resource"
+	}
+}
+
+// buildWinnerAnalysis creates the winner analysis part of the explanation.
+func (e *Explainer) buildWinnerAnalysis(decision *v1alpha1.Decision) string {
+	result := decision.Status.Result
+	if result == nil || result.TargetHost == nil {
+		return ""
+	}
+
+	targetHost := *result.TargetHost
+
+	// Get target host score
+	targetScore := 0.0
+	if result.AggregatedOutWeights != nil {
+		if score, exists := result.AggregatedOutWeights[targetHost]; exists {
+			targetScore = score
+		}
+	}
+
+	// Count hosts evaluated
+	hostsEvaluated := len(result.OrderedHosts)
+	if hostsEvaluated == 0 && result.AggregatedOutWeights != nil {
+		hostsEvaluated = len(result.AggregatedOutWeights)
+	}
+
+	// Calculate score gap to second place
+	gap := e.calculateScoreGap(result.AggregatedOutWeights)
+
+	if gap == 0.0 {
+		// Only one host or perfect score
+		return fmt.Sprintf("Selected: %s (score: %.2f), %d hosts evaluated.",
+			targetHost, targetScore, hostsEvaluated)
+	}
+
+	return fmt.Sprintf("Selected: %s (score: %.2f), gap to 2nd: %.2f, %d hosts evaluated.",
+		targetHost, targetScore, gap, hostsEvaluated)
+}
+
+// calculateScoreGap calculates the gap between first and second place.
+func (e *Explainer) calculateScoreGap(weights map[string]float64) float64 {
+	if weights == nil || len(weights) < 2 {
```

Comment:
While `weights` can be a nil-map, the len() over a nil map is defined as 0. Meaning the nil check is obsolete